articles/multifactor-authentication/custom.md

```diff
@@ -0,0 +1,150 @@
+
+```JS
+function (user, context, callback) {
+
+  if (conditionIsMet()){
+    context.multifactor = {
+      allowRememberBrowser: false,
+      provider: 'any'
+    };
+  }
+
+  callback(null, user, context);
+}
+```
+
+::: note
+Some older rules may use the field `ignoreCookie` here. While deprecated, that field will still function as expected, and will force multi-factor authentication at every login.
+:::
+
+### Access from an extranet
+
+You can have Auth0 request MFA from users whose requests originate from outside the corporate network:
+
+```js
+function (user, context, callback) {
+  var ipaddr = require('ipaddr.js');
+  var corp_network = "192.168.1.134/26";
+
+  var current_ip = ipaddr.parse(context.request.ip);
+  if (!current_ip.match(ipaddr.parseCIDR(corp_network))) {
+    context.multifactor = {
+        provider: 'any',
+        allowRememberBrowser: false
+    };
+  }
+
+  callback(null, user, context);
+}
+```
+
+## Use a custom MFA service
+
+If you are using an MFA provider that does not have Auth0 built-in support or if you are using a service you have created, you can use the [redirect](/rules/redirect) protocol for the integration.
+
+By using the redirect protocol, you interrupt the authentication transaction and redirect the user to a specified URL where they are asked for MFA. If authentication is successful, Auth0 will continue processing the request.
+
+Some MFA options you can implement using the redirect protocol include:
+
+* A one-time code sent via SMS
+* Integration with specialized providers, such as those that require hardware tokens
+
+To use the redirect protocol, edit the `URL` field:
+
+```JS
+function (user, context, callback) {
+
+  if (condition() && context.protocol !== 'redirect-callback'){
+    context.redirect = {
+      url: 'https://your_custom_mfa'
+    };
+  }
+
+  if (context.protocol === 'redirect-callback'){
+    //TODO: handle the result of the MFA step
+  }
+
+  callback(null, user, context);
+}
+```
+
+## Additional Notes
+
+* A tutorial is available on using MFA with the [Resource Owner](/api-auth/tutorials/multifactor-resource-owner-password) endpoint.
+* If you are using MFA for database connections that use Popup Mode, and you are using [Lock](/libraries/lock/v11) or [Auth0.js](/libraries/auth0.js/v9), set `sso` to `true` when defining the options object. If you fail to do this, users will be able to log in without MFA.
+* If you are using MFA after an authentication with one or more social providers, you may need to use your own application `ID` and `Secret` in the connection to the provider's site in place of the default Auth0 development credentials. For instructions on how to get the credentials for each social provider, select your particular from the list at: [Identity Providers](/identityproviders). In production usage, you should always use your own credentials instead of [Auth0 devkeys](/connections/social/devkeys).
```

Comment:
Removed 'may'
```suggestion
* If you are using MFA after an authentication with one or more social providers, you need to use your own application `ID` and `Secret` in the connection to the provider's site in place of the default Auth0 development credentials. For instructions on how to get the credentials for each social provider, select your particular from the list at: [Identity Providers](/identityproviders). In production usage, you should always use your own credentials instead of [Auth0 devkeys](/connections/social/devkeys).
```